torch2trt/torch2trt.py

```diff
@@ -865,3 +896,177 @@ def set_layer_precision(ctx, layer):
+        return result
+
+    def __radd__(self, x):
+        if not isinstance(x, IntWrapper):
+            x = IntWrapper(x)
+        ctx = get_conversion_context()
+        result = IntWrapper(_int_add(x, self))
+        result._raw_trt = ctx.network._network.add_elementwise(x._trt, self._trt, trt.ElementWiseOperation.SUM).get_output(0)
+        return result
+
+    def __rsub__(self, x):
+        if not isinstance(x, IntWrapper):
+            x = IntWrapper(x)
+        ctx = get_conversion_context()
+        result = IntWrapper(_int_sub(x, self))
+        result._raw_trt = ctx.network._network.add_elementwise(x._trt, self._trt, trt.ElementWiseOperation.SUB).get_output(0)
+        return result
+
+    def __rfloordiv__(self, x):
+        if not isinstance(x, IntWrapper):
+            x = IntWrapper(x)
+        ctx = get_conversion_context()
+        result = IntWrapper(_int_floordiv(x, self))
+        result._raw_trt = ctx.network._network.add_elementwise(x._trt, self._trt, trt.ElementWiseOperation.FLOOR_DIV).get_output(0)
+        return result
+
+    def __int__(self):
+        return self
+
+def make_int_wrapper(x):
+    if isinstance(x, IntWrapper):
+        return x
+    else:
+        return IntWrapper(x)
+
+class SizeWrapper(tuple):
+
+    @property
+    def _trt(self):
+        if not hasattr(self, '_raw_trt'):
+            ctx = get_conversion_context()
+            self._raw_trt = ctx.network._network.add_concatenation([d._trt for d in self]).get_output(0)
+        return self._raw_trt
+
+    def __tuple__(self):
+        return self
+
+
+def wrap_ints(x):
+    for y in x:
+        yield make_int_wrapper(y)
+
+
+def make_size_wrapper(args):
+    return SizeWrapper(wrap_ints(args))
+
+
+_original_size = torch.Tensor.size
+_original_getattr = torch.Tensor.__getattribute__
+
+
+def _size_wrapper(input, dim=None):
+
+    if not hasattr(input, '_trt'):
+        if dim is not None:
+            return _original_size(input, dim)
+        else:
+            return _original_size(input)
+
+    ctx = get_conversion_context()
+
+    output = _original_size(input)
+
+    output = make_size_wrapper(output)
+
+    shape_trt = ctx.network._network.add_shape(input._trt).get_output(0)
+
+    for i, d in enumerate(output):
+        d._raw_trt = ctx.network._network.add_slice(shape_trt, [i], [1], [1]).get_output(0)
+
+    if dim is not None:
+        output = output[dim]
+
+    return output
+
+
+_old_getattr = torch.Tensor.__getattribute__
+
+def _new_getattr(self, name):
+    if name == 'shape' and use_shape_wrapping.stack[0]:
+        return _size_wrapper(self)
+    else:
+        return _old_getattr(self, name)
+
+class use_shape_wrapping:
+
+    stack = [True] # default true
+
+    def __init__(self, value: bool):
+        self._value = value
+    
+    def __enter__(self, *args, **kwargs):
+        self.stack.insert(0, self._value)
+
+    def __exit__(self, *args, **kwargs):
+        self.stack.pop()
```

Comment:
What's the reason for tracking the shape wrapping here? I see this is called in several of the helper functions. I'm guessing it's to prevent calling `_size_wrapper` within nested (or recursive) calls to the helper functions, but I don't think I've fully grokked the use case yet. 

Also, should this actually be renamed to a queue? It looks like we're inserting to the front but popping from the back.

Reply:
The purpose is like you describe.  For utility functions we want to access the shape of the torch tensor without modifying the TensorRT network (like in utility functions).

This context manager enables / disables the shape tracking.

This should be a stack, that's a bug! Insert should be append.  I'll update this and test.